filecoin-proofs/src/api/mod.rs

```diff
@@ -535,106 +524,92 @@ pub unsafe extern "C" fn get_staged_sectors(
     raw_ptr(response)
 }
 
-#[repr(C)]
-pub struct FFISectorClass {
-    sector_size: FFISectorSize,
-    seal_proof_partitions: FFISealProofPartitions,
-    post_proof_partitions: FFIPoStProofPartitions,
-}
-
-#[allow(non_camel_case_types)]
-#[derive(Clone, Debug, Copy)]
-#[repr(C)]
-pub enum FFISectorSize {
-    SSB_OneKiB,
-    SSB_TwoHundredFiftySixMiB,
-}
-
-#[allow(non_camel_case_types)]
-#[repr(C)]
-pub enum FFISealProofPartitions {
-    SPP_Two,
-}
-
-#[allow(non_camel_case_types)]
-#[repr(C)]
-pub enum FFIPoStProofPartitions {
-    PPP_One,
+unsafe fn try_into_post_proofs_bytes(
+    proof_partitions: u8,
+    flattened_proofs_ptr: *const u8,
+    flattened_proofs_len: libc::size_t,
+) -> crate::error::Result<Vec<Vec<u8>>> {
+    post_proof_partitions::try_from_u8(proof_partitions)
+        .map(PoStProofBytesAmount::from)
+        .map(usize::from)
+        .map(|chunk| into_proof_vecs(chunk, flattened_proofs_ptr, flattened_proofs_len))
 }
 
-impl From<FFISectorSize> for SectorSize {
-    fn from(x: FFISectorSize) -> Self {
-        match x {
-            FFISectorSize::SSB_OneKiB => SectorSize::OneKiB,
-            FFISectorSize::SSB_TwoHundredFiftySixMiB => SectorSize::TwoHundredFiftySixMiB,
-        }
-    }
+unsafe fn try_into_porep_proof_bytes(
+    proof_partitions: u8,
+    flattened_proofs_ptr: *const u8,
+    flattened_proofs_len: libc::size_t,
+) -> crate::error::Result<Vec<u8>> {
+    porep_proof_partitions::try_from_u8(proof_partitions)
+        .map(PoRepProofBytesAmount::from)
+        .map(usize::from)
+        .map(|chunk| into_proof_vecs(chunk, flattened_proofs_ptr, flattened_proofs_len))
+        .and_then(|vs| {
```

Comment:
> If I'm reading this right, we sequentially apply a chain of three function calls in order to find out the number of partitions (which could/should have been calculated by the caller, so doesn't provide new information) — but then only use that number to break the proofs into individual proofs.

You are not reading this right. The caller provides the number of proof partitions N as a u8, and then the dynamic array which the caller provides is chunked using that value. The `porep_proof_partitions::try_from_u8` function enforces an invariant, which is that there's only a few values of N which our system can handle.

The use case is more compelling when accepting a dynamic array of flattened PoSt proofs, but validating / chunking PoRep proofs in the same way seems sane to me.

Anyways, I'm sorry that you don't like all the types. I don't agree that the types have "no clear purpose." Quite the contrary; they allow us to enforce invariants at compile time.

The good news is that I'm going to delete almost all of those types in a coming PR, because of backwards-compatibility issues that we discussed last week.